packages/flipper-plugin-react-query-native-devtools/src/index.tsx

```diff
@@ -1,73 +1,179 @@
+  handleQueryRefetch: (query: ExtendedQuery) => void;
+  handleQueryRemove: (query: ExtendedQuery) => void;
+};
 
-  const onQueryRefetch = (query: Query): void => {
-    client.call('query:refetch', { queryHash: query.queryHash });
-  };
+export type ExtendedQuery = Query & {
+  status: QueryStatus;
+  dataUpdateCount: number;
+  observersCount: number;
+  isQueryActive: boolean;
+};
 
-  const onQueryRemove = (query: Query): void => {
-    client.call('query:remove', { queryHash: query.queryHash });
-  };
+const extendQuery = (query: Query): ExtendedQuery => {
+  const extendedQuery = query as ExtendedQuery;
+  extendedQuery.status = query.state.status;
+  extendedQuery.dataUpdateCount = query.state.dataUpdateCount;
+  extendedQuery.observersCount = getObserversCounter(query);
+  extendedQuery.isQueryActive = isQueryActive(query);
 
-  return (
-    <Container>
-      <QueryTable queries={Object.values(queries)} onSelect={setSelectedQueries} />
-      <Sidebar query={query} onQueryRefetch={onQueryRefetch} onQueryRemove={onQueryRemove} />
-    </Container>
-  );
+  return extendedQuery;
 };
+export const plugin = (client: PluginClient<Events, Methods>): PluginReturn => {
+  const queries = createDataSource<ExtendedQuery, 'queryHash'>([], {
+    key: 'queryHash',
+  });
+  const selectedQueryId = createState<string | undefined>(undefined);
 
-type Payload = {
-  queries: string;
-};
+  client.onMessage('queries', (event) => {
+    // That happens onConnect only
+    queries.clear();
+    selectedQueryId.set(undefined);
 
-export default class ReactQueryDevtoolsFlipperPlugin extends FlipperPlugin<{}, any, PersistedState> {
-  static title = 'React Query Devtools';
-  static icon = 'app-react';
-  static defaultPersistedState = {
-    queries: [],
-  };
-  static persistedStateReducer(persistedState: PersistedState, method: string, payload: Payload): PersistedState {
-    if (method === 'queries') {
-      return {
-        ...persistedState,
-        queries: parse(payload.queries) as Query[],
-      };
+    parse(event.queries).forEach((query: ExtendedQuery) => {
+      queries.append(extendQuery(query));
+    });
+  });
+
+  client.onMessage('queryCacheEvent', (event) => {
+    const cashEvent = parse(event.cashEvent) as QueryCacheNotifyEvent;
+    const {
+      type,
+      query,
+      query: { queryHash },
+    } = cashEvent;
+
+    if (!type) {
+      return;
+    }
+
+    switch (type) {
+      case 'queryAdded':
+        queries.append(extendQuery(query));
+        break;
+      case 'queryRemoved':
+        queries.deleteByKey(queryHash);
+        break;
+      case 'queryUpdated':
+      case 'observerAdded':
+      case 'observerRemoved':
+      case 'observerResultsUpdated':
+        queries.upsert(extendQuery(query));
+
+        // To re-render the sidebar when we have selected query and the query is updated
+        if (selectedQueryId.get()?.slice(11) === queryHash) {
+          selectedQueryId.set(makeQuerySelectionKey(query));
```

Comment:
We need to change `selectedQueryId` when it's updated. The comparison happens by value. Even I tried to undefined it and set some intermediate values to selectedQueryId at the end of the day React part does not react on it. The final result set here is compared by value is QuerySidebar and if the value is the same than we have no re-render. I tried to strike re-render with standard React ways: by changing props, state and children but it did Not bring fruits. Maybe we can come up with the better solution, but this one works Ok.